From 750c3526cf7cf7ffce22cfce9376bb78c9639df6 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 12:58:43 -0400
Subject: [ethernet] qlge: Do not propaged vlan tag offloads to vlans

Message-id: <1408625927-3701-13-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88531
O-Subject: [PATCH RHEL 7.1 12/16] qlge: Do not propaged vlan tag offloads to vlans
Bugzilla: 1089357
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Vlad Yasevich <vyasevic@redhat.com>

Bug 1089357 (https://bugzilla.redhat.com/show_bug.cgi?id=1089357)

net commit id f6d1ac4b5f15f57929fe0fa283b3a45dfec717a0

qlge driver turns off NETIF_F_HW_CTAG_FILTER, but forgets to
turn off HW_CTAG_TX and HW_CTAG_RX on vlan devices.  With the
current settings, q-in-q will only generate a single vlan header.
Remember to mask off CTAG_TX and CTAG_RX features in vlan_features.

CC: Shahed Shaikh <shahed.shaikh@qlogic.com>
CC: Jitendra Kalsaria <jitendra.kalsaria@qlogic.com>
CC: Ron Mercer <ron.mercer@qlogic.com>
Signed-off-by: Vlad Yasevich <vyasevic@redhat.com>
Acked-by: Jitendra Kalsaria <jitendra.kalsaria@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlge/qlge_main.c b/drivers/net/ethernet/qlogic/qlge/qlge_main.c
index 777fab3..2d8d05d 100644
--- a/drivers/net/ethernet/qlogic/qlge/qlge_main.c
+++ b/drivers/net/ethernet/qlogic/qlge/qlge_main.c
@@ -4766,7 +4766,9 @@ static int qlge_probe(struct pci_dev *pdev,
  ndev->features = ndev->hw_features;
  ndev->vlan_features = ndev->hw_features;
  /* vlan gets same features (except vlan filter) */
- ndev->vlan_features &= ~NETIF_F_HW_VLAN_CTAG_FILTER;
+ ndev->vlan_features &= ~(NETIF_F_HW_VLAN_CTAG_FILTER |
+     NETIF_F_HW_VLAN_CTAG_TX |
+     NETIF_F_HW_VLAN_CTAG_RX);
 
  if (test_bit(QL_DMA64, &qdev->flags))
   ndev->features |= NETIF_F_HIGHDMA;
-- 
1.7.1